test/unit/cmap/handshake/client_metadata.test.ts

```diff
@@ -1,149 +1,511 @@
+        }
+      ],
+      azure: [
+        {
+          context: 'no additional metadata',
+          env: [['FUNCTIONS_WORKER_RUNTIME', 'non-empty']],
+          outcome: {
+            name: 'azure.func'
+          }
+        }
+      ],
+      gcp: [
+        {
+          context: 'no additional metadata',
+          env: [['FUNCTION_NAME', 'non-empty']],
+          outcome: {
+            name: 'gcp.func'
+          }
+        },
+        {
+          context: 'FUNCTION_MEMORY_MB provided',
+          env: [
+            ['FUNCTION_NAME', 'non-empty'],
+            ['FUNCTION_MEMORY_MB', '1024']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            memory_mb: new Int32(1024)
+          }
+        },
+        {
+          context: 'FUNCTION_REGION provided',
+          env: [
+            ['FUNCTION_NAME', 'non-empty'],
+            ['FUNCTION_REGION', 'region']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            region: 'region'
+          }
+        },
+        {
+          context: 'FUNCTION_TIMEOUT_SEC provided',
+          env: [
+            ['FUNCTION_NAME', 'non-empty'],
+            ['FUNCTION_TIMEOUT_SEC', '12345']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            timeout_sec: new Int32(12345)
+          }
+        }
+      ],
+      vercel: [
+        {
+          context: 'no additional metadata',
+          env: [['VERCEL', 'non-empty']],
+          outcome: {
+            name: 'vercel'
+          }
+        },
+        {
+          context: 'VERCEL_URL provided',
+          env: [
+            ['VERCEL', 'non-empty'],
+            ['VERCEL_URL', 'provided-url']
+          ],
+          outcome: {
+            name: 'vercel',
+            url: 'provided-url'
+          }
+        },
+        {
+          context: 'VERCEL_REGION provided',
+          env: [
+            ['VERCEL', 'non-empty'],
+            ['VERCEL_REGION', 'region']
+          ],
+          outcome: {
+            name: 'vercel',
+            region: 'region'
+          }
+        }
+      ]
+    };
+
+    for (const [provider, testsForEnv] of Object.entries(tests)) {
+      context(provider, () => {
+        for (const { context, env: faasVariables, outcome } of testsForEnv) {
+          const setupEnv = () => {
+            for (const [k, v] of faasVariables) {
+              if (v != null) {
+                process.env[k] = v;
+              }
+            }
+          };
+
+          const cleanupEnv = () => {
+            for (const [k] of faasVariables) {
+              delete process.env[k];
+            }
+          };
+
+          it(context, () => {
+            try {
+              setupEnv();
+              const { env } = makeClientMetadata({ driverInfo: {} });
+              expect(env).to.deep.equal(outcome);
+            } finally {
+              cleanupEnv();
+            }
+          });
+
+          it('always places name as the last key', () => {
```

Comment:
here's that test for name being last, so if any new faas envs are added we'll keep this up